examples/bubble_example.py

```diff
@@ -0,0 +1,237 @@
+        agent_builder=DummyAgent,
+    ),
+)
+
+# Locations from inspecting the `map.net.xml` using netedit
+bubble_locations = {
+    "i80": [
+        Bubble(
+            zone=PositionalZone(pos=(0, 0), size=(20, 100)),
+            actor=SocialAgentActor(
+                name="dummy0", agent_locator="examples:dummy_agent-v0"
+            ),
+            follow_vehicle_id="history-vehicle-314",
+            exclusion_prefixes=("history-vehicle-314",),
+            follow_offset=(0, 0),
+            margin=10
+        ),
+    ],
+    "peachtree": [
+        Bubble(
+            zone=MapZone(start=("E9", 0, 5), length=40, n_lanes=2),
+            actor=SocialAgentActor(
+                name="dummy0", agent_locator="examples:dummy_agent-v0"
+            ),
+        )
+    ],
+    "us101": [
+        Bubble(
+            zone=MapZone(start=("gneE01.132", 0, 1), length=120, n_lanes=5),
+            actor=SocialAgentActor(
+                name="dummy0", agent_locator="examples:dummy_agent-v0"
+            ),
+        )
+    ]
+}
+
+
+def main(
+    script: str,
+    ngsim_example: str,
+    headless: bool,
+    seed: int,
+    episodes: int,
+    start_time: float,
+    run_time: float,
+):
+    logger = logging.getLogger(script)
+    logger.setLevel(logging.INFO)
+
+    logger.debug("initializing SMARTS")
+
+    timestep = 0.1
+    run_steps = int(run_time / timestep)
+    smarts = SMARTS(
+        agent_interfaces={},
+        traffic_sim=None,
+        envision=None if headless else Envision(),
+        fixed_timestep_sec=timestep,
+    )
+    random_seed(seed)
+
+    scenario = str(Path(__file__).parent.parent / "scenarios/NGSIM" / ngsim_example)
+    scenario_list = Scenario.get_scenario_list([scenario])
+    scenarios_iterator = Scenario.variations_for_all_scenario_roots(scenario_list, [])
+
+    class obs_ref:
+        last_observations: Dict[str, Observation] = None
+
+    def observation_callback(obs: Observation):
+        obs_ref.last_observations = obs
+
+    for scenario in scenarios_iterator:
+        scenario.bubbles.clear()
+        scenario.bubbles.extend(bubble_locations[ngsim_example])
+
+        # XXX replace with AgentSpec appropriate for IL model
+        agent_manager: AgentManager = smarts.agent_manager
+        agent_manager.add_social_observation_callback(
+            observation_callback, "bubble_watcher"
+        )
+
+        for episode in range(episodes):
+            logger.info(f"starting episode {episode}...")
+            smarts.reset(scenario, start_time=start_time)
+            traffic_history_provider: TrafficHistoryProvider = smarts.get_provider_by_type(TrafficHistoryProvider)
+            used_history_ids = set()
+
+            bubble_manager: BubbleManager = smarts._bubble_manager
+            bubbles = bubble_manager.bubbles
+            agent = UsedAgent(logger=logger)
```

Comment:
It seems that the `agent` here is not used in this example. And I have some questions:

1. What is the vehicle being followed controlled by? (whose vehicle id="follow_vehicle_id"). Does it serve as a `social agent` or a `ego vehicle`？
2. If the vehicle being followed is a ego vehicle, should I call `scenario.set_ego_missions` and `smarts.switch_ego_agents`?

Reply:
I will try and answer the questions.

> It seems that the `agent` here is not used in this example.

Yes and no, in this case we are operating with bubble captured vehicles. These cannot come through the normal smarts interface. This is more a mock-up of the utility but the interface needs to be resolved. There is a parallel effort ongoing for work with traffic https://github.com/huawei-noah/SMARTS/pull/1440 that will help with resolving that interface.

The agent is used here in a different way than before by overriding a social agent action:
https://github.com/huawei-noah/SMARTS/blob/55453bb77f53a35e1829ac44d8912f4519559809/examples/bubble_example.py#L167-L169

If the logging is turned on you get the following:
```bash
INFO:history-vehicles-replacement-example:starting episode 0...
INFO:history-vehicles-replacement-example:Agent act called first time.
INFO:history-vehicles-replacement-example:starting episode 1...
INFO:history-vehicles-replacement-example:Agent act called first time.
```

relating to:
https://github.com/huawei-noah/SMARTS/blob/55453bb77f53a35e1829ac44d8912f4519559809/examples/bubble_example.py#L53-L61

> 1. What is the vehicle being followed controlled by? (whose vehicle id="follow_vehicle_id"). Does it serve as a `social agent` or a `ego vehicle`？

There are two cases for this:
  - If using `follow_vehicle_id` this bubble will follow the vehicle with the given id. The followed vehicle could be any vehicle in the dataset regardless of if it is controlled by an `agent` or simply a data replay vehicle.
  - If using `follow_actor_id` this bubble will follow the agent with that ID when it shows up in the simulation.

The following was to prevent specific vehicles from being taken over by a social vehicle:
```python
exclusion_prefixes=("history-vehicle-314",),
```

> 2. If the vehicle being followed is a ego vehicle, should I call `scenario.set_ego_missions` and `smarts.switch_ego_agents`?

That would be the case at current. Using the `exclusion_prefixes` you can ensure that the vehicle is not already controlled by a social agent agent.

assuming you are taking over vehicles in the dataset using `scenario.set_ego_missions` & `smarts.switch_ego_agents` 

> 3. If the followed vehicle's  agent id is not provided by user (by `AgentInterface`) , we do not know its agent id, so it is unrealistic to define a bubble in advance.

I am not quite what to respond to since this is not a question but there is a design issue here that requires some explanation.
  - The current bubbles are defined in advance but adding them into the scenario dynamically during simulation is not yet supported (but easily implemented.) But since you would be considering making modifications to the sim state before reset (i.e. `scenario.set_ego_missions` & `smarts.switch_ego_agents`) it is possible to set the bubbles in the scenario at the same time.
```python
# TODO: interface to be improved.
scenario.bubbles.clear()
scenario.bubbles.extend([Bubble(..)])
scenario.set_ego_missions(..)
...
smarts.switch_ego_agents(..)
...
smarts.reset(scenario, ..)
```
  - Bubbles acquire vehicles for social agents at runtime. The idea experimented here is that these social agents can be controlled because bubbles were not originally intended to capture ego vehicles.
  - The capture of ego vehicles could result in `smarts.reset(..)` and `smarts.step(..)` having to generate new ids, or needing a supply of ids for the bubble to assign by injecting bubble information into the agent interface.

I do not yet understand what your intention is yet for use of the travelling bubble. Part of what this rough implementation is meant to uncover your intentions. I have some of my own questions:

- Do you need the moving bubble to follow an ego agent previously added by `scenario.set_ego_missions` & `smarts.switch_ego_agents`?
- Is your intention that bubble just provides information about vehicles entering/exiting without automated takeover of those vehicles by social agents?
  - You would want to find vehicles in the bubble?
  - Would you want to manage take-over of vehicles directly?
- Do you need to know the agent ids from the bubble before-hand or are dynamic ids (agent ids provided by smarts) reasonable?

Reply:
@Gamenot Thanks for the replay.

> - Do you need the moving bubble to follow an ego agent previously added by scenario.set_ego_missions & smarts.switch_ego_agents?

Yes, exactly.  What I need is just a interface like this,
```python
start_time = 50
scenario = Scenario(...)
smarts = SMARTS()
ego_vehicle_id = "vehicle-10"
ego_agent_id = "agent-10"

ego_missions = {ego_agent_id: get_vehicle_mission(ego_vehicle_id, start_time)}
agent_interfaces = {ego_agent_id: AgentInterface(...)}
scenario.set_ego_missions(ego_missions)
smarts.switch_ego_agents(agent_interfaces)
scenario.bubbles.clear()
scenario.bubbles.extend([Bubble(follow_actor_id=ego_agent_id)])
# or, scenario.bubbles.extend([Bubble(follow_vehicle_id=ego_vehicle_id)])
agent_manager: AgentManager = smarts.agent_manager
agent_manager.add_social_observation_callback(
    obs_state.observation_callback, "bubble_watcher"
)

class EgoAgent:
    def act(self, observation):
        ...
        return action

class SocialAgent:
    def act(self, observation):
        ...
        return action

ego_agent = EgoAgent()
social_agent = SocialAgent()
ego_observations = smarts.reset(scenario, start_time)
...
for social_agent_id in obs_state.last_observations:
        agent_manager.reserve_social_agent_action(
            social_agent_id, social_agent.act(obs_state.last_observations[agent_id])
        )
ego_action = {ego_agent_id: ego_agent.act(obs) for ego_agent_id, obs in ego_observations.items()}
smarts.step(ego_action)
```

> Is your intention that bubble just provides information about vehicles entering/exiting without automated takeover of those vehicles by social agents?
> - You would want to find vehicles in the bubble?
> - Would you want to manage take-over of vehicles directly?

I do not need to take-over vehicles by myself now.

> Do you need to know the agent ids from the bubble before-hand or are dynamic ids (agent ids provided by smarts) reasonable?

I want to specify the ego_vehicle's (vehicle to be followed) agent id, by I don't need to pre-define social agents' agent ids (because the vehicles in bubble are dynamically changing).